### PR TITLE
feat(artifact): add artifacts CRUDL

### PR DESCRIPTION
**Pull request by xjacka**
https://github.com/i-am-bee/internal/issues/18

### CODE REVIEW

File: src/artifacts/dtos/artifact-shared.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+export const artifactSharedSchema = {
+  type: 'object',
+  required: ['id', 'type', 'created_at', 'object', 'metadata', 'name', 'description'],
+  properties: {
+    id: { type: 'string' },
+    object: { const: 'artifact.shared' },
+    metadata: metadataSchema,
+    created_at: { type: 'number' },
+    name: { type: 'string' },
+    description: { type: 'string' }
```
Comment (pilartomas) [inline src/artifacts/dtos/artifact-shared.ts:32]: nullable

File: src/artifacts/entities/artifact.entity.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  @ManyToOne()
+  thread: Ref<Thread>;
+
+  @ManyToOne()
+  message: Ref<Message>;
+
+  @Unique()
+  @Property()
+  accessSecret?: string;
```
Comment (pilartomas) [inline src/artifacts/entities/artifact.entity.ts:48]: remove unique? we identify artifacts by ID.

File: src/artifacts/dtos/artifact-create.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+import { FromSchema, JSONSchema } from 'json-schema-to-ts';
+
+import { ArtifactType } from '../entities/artifact.entity.js';
+
+import { artifactSchema } from './artifact.js';
+
+import { metadataSchema } from '@/schema.js';
+
+const commonartifactProperties = {
```
Comment (pilartomas) [inline src/artifacts/dtos/artifact-create.ts:25]: typo -> `commonArt..`

File: src/artifacts/artifacts.service.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        source_code: (artifact as AppArtifact).sourceCode
+      };
+  }
+}
+
+function getSecret() {
+  return crypto.randomBytes(24).toString('base64url');
+}
+
+export async function createArtifact(body: ArtifactCreateBody): Promise<ArtifactCreateResponse> {
```
Comment (pilartomas) [inline src/artifacts/artifacts.service.ts:85]: Do we want to allow users to create artifacts? Shouldn't artifacts come from within the system?

File: src/artifacts/artifacts.service.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  const dto = {
+    id: artifact.id,
+    object: 'artifact.shared' as const,
+    type: artifact.type,
+    metadata: artifact.metadata ?? {},
+    created_at: dayjs(artifact.createdAt).unix(),
+    source_code: (artifact as AppArtifact).sourceCode,
+    share_url: artifact.accessSecret
+      ? `/v1/artifacts/${artifact.id}/shared?secret=${artifact.accessSecret}`
+      : '',
```
Comment (pilartomas) [inline src/artifacts/artifacts.service.ts:68]: I'd put this into `toDto`

Comment (pilartomas) [reply src/artifacts/artifacts.service.ts:68]: Fallback should be `null`

File: src/artifacts/artifacts.service.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    name: artifact.name,
+    description: artifact.description ?? ''
```
Comment (pilartomas) [inline src/artifacts/artifacts.service.ts:70]: fallback should be `null`

File: src/artifacts/artifacts.service.ts [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  return crypto.randomBytes(24).toString('base64url');
+}
+
+export async function createArtifact(body: ArtifactCreateBody): Promise<ArtifactCreateResponse> {
+  const thread = await ORM.em.getRepository(Thread).findOneOrFail({ id: body.thread_id });
+  const message = await ORM.em.getRepository(Message).findOneOrFail({ id: body.message_id });
+
+  if (message.thread.id !== thread.id) {
+    throw new APIError({ message: 'Thread message mismatch', code: APIErrorCode.INVALID_INPUT });
+  }
```
Comment (pilartomas) [inline src/artifacts/artifacts.service.ts:91]: Why does user need to specify the thread then?

File: src/artifacts/artifacts.service.ts [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+  if (body.type === ArtifactType.APP) {
```
Comment (pilartomas) [inline src/artifacts/artifacts.service.ts:93]: `switch` might be clearer

File: src/artifacts/artifacts.service.ts (hunk 5 above)
Comment (xjacka) [reply src/artifacts/artifacts.service.ts:68]: Why? This can be in the public DTO.

File: src/artifacts/artifacts.service.ts (hunk 4 above)
Comment (xjacka) [reply src/artifacts/artifacts.service.ts:85]: This is a good question. but the artifact can be part of the message so we should somehow index this part. We can discuss this.

**Review by pilartomas — APPROVED** [review]

**Pull request merged**